### PR TITLE
Added Mmn11

### PR DESCRIPTION
**Pull request by yahelil**
Reviewing MMN11

### CODE REVIEW

File: Maman11/Q1/my_add.c [hunk 1]
```diff
@@ -1,25 +1,10 @@
 #include <stdio.h>
 #include <stdlib.h>
+#define DATA_SIZE 33
```
Comment (yotamlevit) [inline Maman11/Q1/my_add.c]: DATA_SIZE means nothing. what is data. give it a more indicative name

File: Maman11/Q1/my_add.c [hunk 2]
```diff
@@ -1,25 +1,10 @@
...
-    printf("\nYou entered : %d and %d\n\n", num1, num2);
-
-    binary1 = decimalToBinary(num1);
-    binary2 = decimalToBinary(num2);
-    printf("\nThe sum of %d and %d is: %d or %d in decimal\n", num1, num2, addBinary(binary1, binary2), num1+num2);
-    return 0;
-}
-
-int decimalToBinary(int num){
+unsigned int decimalToBinary(int num){
```
Comment (yotamlevit) [inline Maman11/Q1/my_add.c]: This function is wrong.
Converting a number to a string and then to a number is not efficient both in memory and in compute.
Try to convert a number to string and then to return the string.

Think about the number 1025 its binery value is 10000000001 this value is bigger then int.MaxValue that is 2147483647

Thus your program will not work

File: Maman11/Q1/my_add.c [hunk 3]
```diff
@@ -1,25 +1,10 @@
...
     int result = num;
-    char binary[33];
+    char binary[DATA_SIZE];
     int index = 0;
 
     while (result > 0) {
```
Comment (yotamlevit) [inline Maman11/Q1/my_add.c]: I cannot write a review on the code block below but
on line wiht '\0' you should use a defin for it.

Using. define makes the code clearer.
Think about this:
#define EOS '\0' // (EOS = End Of String)

Then in every line you need \0 it will be clearer what is \0

File: Maman11/Q1/my_add.c [hunk 4]
```diff
@@ -44,7 +29,7 @@ int decimalToBinary(int num){
     printf("Binary representation: %s\n", binary);
```
Comment (yotamlevit) [inline Maman11/Q1/my_add.c]: SOLID - Read about SOLID

SOLID is:
Single Responsibility Principle (SRP)
Open-Closed Principle (OCP)
Liskov Substitution Principle (LSP)
Interface Segregation Principle (ISP)
Dependency Inversion Principle (DIP)


You dont need to understand everything at the moment and this is more a principle for OOP
but the important thing is S =  Single Responsibility Principle (SRP)
Each pice of code you write should be a single responsibility.
In this function you Both convert a number to binary and print the number. This is wrong
From solid we understand that each function hsould have only one responsibility thus this function should one hae one responsibility

File: Maman11/Q1/my_add.c (hunk 2 above)
Comment (yahelil) [reply Maman11/Q1/my_add.c]: I have to convert it to unsigned int one way or another since it's required that my_add returns an unsigned int. Does it matter if I do in the my_add or decimalToBinary?

Comment (yotamlevit) [reply Maman11/Q1/my_add.c]: I think you got this a bit wrong. 

my_add - A function that only do a binary addition 

Also you should have a function that converts a number to a string in a binary representation. 
So in other words you have a function that gets a number (type is unsigned int because the program uses unsigned int for numbers in this case)
And returns the output that is the given number`s string in a binary representation. 

in other words 
char* numToBinStr(unsigned int)

File: Maman11/Q1/my_add.c [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+unsigned int my_add(unsigned int a, unsigned int b) {
+    unsigned int sum = 0, carry = 0, result = 0, bitA, bitB, i;
+
+    for (i = 0; i < sizeof(unsigned int) * BYTE_SIZE;i++) {
+        bitA = (a >> i) & 1;
+        bitB = (b >> i) & 1;
+
+        /* Calculate the sum of bits and carry */
+        sum = bitA ^ bitB  ^ carry;
+        printf("\nbitA: %d, bitB: %d, sum: %d", bitA, bitB, sum);
```
Comment (yotamlevit) [inline Maman11/Q1/my_add.c]: Why the print?
Did the question required it?

File: Maman11/Q1/my_add.c [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    return result;
+}
+
+int main() {
+    unsigned int num1, num2, sum;
+    char* binary1, *binary2;
+    printf("Enter a number: ");
+    scanf("%d", &num1); // Reads an integer from the user
```
Comment (yotamlevit) [inline Maman11/Q1/my_add.c]: You should use %u and not %d

File: Maman11/Q1/my_add.c [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+    printf("Enter a second number: ");
+    scanf("%d", &num2); // Reads an integer from the user
```
Comment (yotamlevit) [inline Maman11/Q1/my_add.c]: You should use %u and not %d

File: Maman11/Q1/my_add.c [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    printf("\nYou entered : %d and %d\n\n", num1, num2);
+
+    print_binary(num1);
+    print_binary(num2);
+    sum = my_add(num1, num2);
+    printf("\nThe sum of %d and %d is:\n%d in decimal\n", num1, num2, sum);
```
Comment (yotamlevit) [inline Maman11/Q1/my_add.c]: You should use %u and not %d'
 if you use %d the compiler thinks you use regular int so the range of numbers will be -2^31 -> 2^31

File: Maman11/Q2/my_suffix.c [hunk 9]
```diff
@@ -0,0 +1,45 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#define STRING_LENGTH 127
```
Comment (yotamlevit) [inline Maman11/Q2/my_suffix.c]: Q2 says the input should be max 127 chars not included the \0 so you should have 128

File: Maman11/Q2/my_suffix.c [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    return suffix;
+}
+
+void print_ending(char str[]){
+    printf("%s\n", str);
+}
+
+int main()
+{
+    char c;
```
Comment (yotamlevit) [inline Maman11/Q2/my_suffix.c]: It docent matter but just so you know you can declare both char variables (str and c ) on the same line
char c, str[num]

File: Maman11/Q1/my_add.c [hunk 11]
```diff
@@ -0,0 +1,55 @@
+#include <stdio.h>
+#include <stdlib.h>
+#define BYTE_SIZE 8
+
+void print_binary(int num) {
```
Comment (yotamlevit) [inline Maman11/Q1/my_add.c:5]: Add code documentation for all functions.
Inputs, outputs, how does the function work

this is an example:
/*
 This function finds suffix in a string that start with a given char
 When the function finds the suffix it prints it and counts it


 The function iterates over the string using a for loop.

 param str: (char[]) the string to iterate over
 param c: (char) the char that start the suffix.

 @return: (int) the number of suffix in the string.
          if a suffix is found the program prints it.
*/

I was lazy with the explanation but this is the format

File: Maman11/Q1/my_add.c [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    if (num > 1) {
+        print_binary(num / 2);  // Recursive call for higher bits
```
Comment (yotamlevit) [inline Maman11/Q1/my_add.c]: This type of comments will give you warnings in c90 look on the make file that I did (you should use the flags -Wall -ansi -pedantic)

c90 (old c lang) uses `/* comment */` to comment so just check for warnings

File: Maman11/Q1/my_add.c [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    
+    // Print the least significant bit (LSB)
+    printf("%d", num % 2);
+}
+
+unsigned int my_add(unsigned int a, unsigned int b) {
```
Comment (yotamlevit) [inline Maman11/Q1/my_add.c:21]: Add code documentation for all functions.

File: Maman11/Q2/my_suffix.c [hunk 14]
```diff
@@ -0,0 +1,45 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#define STRING_LENGTH 127
+
+int suffix(char str[], char c){
```
Comment (yotamlevit) [inline Maman11/Q2/my_suffix.c:7]: Add code documentation for all functions.

File: Maman11/Q2/my_suffix.c [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    bool read = false;
```
Comment (yotamlevit) [inline Maman11/Q2/my_suffix.c]: This is good but for uni you cannot use `studbool.h` so try it with a different way. I do want you to have true, false words.

Still this is vary good idea when I code in my job I use the true false of that is included.

Try this and understand the meaning.
`typedef enum {FALSE = 0, TRUE} boolean;`

File: Maman11/Q2/my_suffix.c [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    int suffix = 0, len = strlen(str), i = 0;
+    while (str[i]){
```
Comment (yotamlevit) [inline Maman11/Q2/my_suffix.c]: I want you to iterate over the string in a cleaner way.
In c you can do amazing stuff using pointers.
Try to create. achar* pointer to the string and with this pointer iterate over the string.
DO this using a. char* pointer and a for loop

ALso this about a way to know when you are at the end of a sting. (`\0`)


I will help you more.
str[] is an array of chars in the memory. Each char is a byte. 
So str will look like this in the memory - 0000 (this is the start address): byte, byte. byte .......


So you can create a pointer for the first byte and then iterate over the bytes

File: Maman11/Q2/my_suffix.c [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+        if (c == str[i]){
```
Comment (yotamlevit) [inline Maman11/Q2/my_suffix.c]: Again if you will use a for loop and a cleaner way all the functions would be cleaner and better

File: Maman11/Q2/my_suffix.c [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+            read = true;
+        }
+        if (read){
```
Comment (yotamlevit) [inline Maman11/Q2/my_suffix.c]: If you set read to true you can just do the following:

if (char is equal)
   do the prints and count ++


this can be the only if in the loop

File: Maman11/Q2/my_suffix.c [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+            print_ending(&str[i]);
+            read = false;
+            suffix += 1;
+
+        }
+
+        i++;
+    }
+    return suffix;
```
Comment (yotamlevit) [inline Maman11/Q2/my_suffix.c]: suffix is not a good variable name for this. 
This is a counter for the number of suffixes that the string has

File: Maman11/Q2/my_suffix.c [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+void print_ending(char str[]){
```
Comment (yotamlevit) [inline Maman11/Q2/my_suffix.c]: This function is useless. 
YOu can do this print where you need to pring

File: Maman11/Q2/my_suffix.c [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+{
+    char c;
+    char str[STRING_LENGTH];
+    int num;
+    printf("Enter a char: ");
+    scanf(" %c", &c);
+    printf("Enter a string: ");
+    scanf("%s", &str);
+    printf("The character is %c and the string is %s\n", c, str);
+    num = suffix(str, c);
```
Comment (yotamlevit) [inline Maman11/Q2/my_suffix.c]: num is not a good name

File: Maman11/Q2/my_suffix.c [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+int main()
+{
+    char c;
+    char str[STRING_LENGTH];
+    int num;
+    printf("Enter a char: ");
+    scanf(" %c", &c);
+    printf("Enter a string: ");
+    scanf("%s", &str);
+    printf("The character is %c and the string is %s\n", c, str);
```
Comment (yotamlevit) [inline Maman11/Q2/my_suffix.c]: In general add comments and documentation for all the functions

File: Maman11/Q1/my_add.c (hunk 5 above)
Comment (yahelil) [reply Maman11/Q1/my_add.c]: No, used it to check something and forgot to delete
